src/linkediting.js

```diff
@@ -55,5 +60,90 @@ export default class LinkEditing extends Plugin {
 
 		// Enable two-step caret movement for `linkHref` attribute.
 		bindTwoStepCaretToAttribute( editor.editing.view, editor.model, this, 'linkHref' );
+
+		// Setup highlight over selected link.
+		this._setupLinkHighlight();
+	}
+
+	/**
+	 * Adds highlight over link which has selection inside, together with two-step caret movement indicates whenever
+	 * user is typing inside the link.
+	 *
+	 * @private
+	 */
+	_setupLinkHighlight() {
+		const editor = this.editor;
+		const model = this.editor.model;
+		const doc = model.document;
+		const highlightDescriptor = {
+			id: 'linkBoundaries',
+			class: 'ck-link_selected',
+			priority: 1
+		};
+
+		// Convert linkBoundaries marker to view highlight.
+		editor.conversion.for( 'editingDowncast' )
+			.add( downcastMarkerToHighlight( {
+				model: 'linkBoundaries',
+				view: highlightDescriptor
+			} ) );
+
+		// Create marker over whole link when selection has "linkHref" attribute.
+		doc.on( 'change', () => {
+			const selection = doc.selection;
+
+			// Create marker over link when selection is inside or remove marker otherwise.
+			if ( selection.hasAttribute( 'linkHref' ) ) {
+				const modelRange = findLinkRange( selection.getFirstPosition(), selection.getAttribute( 'linkHref' ) );
+				const marker = model.markers.get( 'linkBoundaries' );
+
+				if ( !marker || !marker.getRange().isEqual( modelRange ) ) {
+					model.change( writer => {
+						writer.setMarker( 'linkBoundaries', modelRange );
+					} );
+				}
+			} else if ( model.markers.has( 'linkBoundaries' ) ) {
+				model.change( writer => {
+					writer.removeMarker( 'linkBoundaries' );
+				} );
+			}
+		} );
+
+		// Custom converter for selection's "linkHref" attribute - when collapsed selection has this attribute it is
+		// wrapped with <span> similar to that used by highlighting mechanism. This <span> will be merged together with
+		// highlight wrapper. This prevents link splitting When selection is at the beginning or at the end of the link.
+		// Without this converter:
+		//
+		//		<a href="url">{}</a><span class="ck-link_selected"><a href="url">foo</a></span>
+		//
+		// When converter is applied:
+		//
+		//		<span class="ck-link_selected"><a href="url">{}foo</a></span>
+		editor.editing.downcastDispatcher.on( 'attribute:linkHref', ( evt, data, conversionApi ) => {
```

Comment:
Honestly, I don't feel good about this solution. It seems overcomplicated compared to the other one (changing the highlight priority to higher than `<a>` element priority).

You provide a new, custom, nontrivial converter just to wrap selection in an edge case.

Instead, we should have a discussion whether model selection should be treated as "inside" marker or "outside" marker when it is on a marker's boundary. Maybe this should be somehow configurable?

Reply:
Okay, after some F2F discussion I'd conclude that:

* let's leave this solution as it is,
* let's have a discussion about how markers and selection should work together,
* let's have a discussion whether this kind of "converter" should be done as a helper, maybe even as a part of current model-to-view markers converters.